<a href="https://colab.research.google.com/github/Yanina2105/Heart-attack_project/blob/develop/Limpieza_de_datos_2020.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploración del dataset**

## **Cargar librerias y revisión básica**

In [ ]:
from pathlib import Path

import numpy as np
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
DATA_PATH = Path('/content/drive/MyDrive/Proyecto_de_tesis/datos_2020/data')
PROCESSED_DATA_PATH = DATA_PATH / 'processed'

FINAL_FILE_WITH_NANS_PATH = PROCESSED_DATA_PATH / 'heart_with_nans_2020.csv'

In [ ]:
heart_2020_df = pd.read_csv(FINAL_FILE_WITH_NANS_PATH)

/tmp/ipython-input-2172687843.py:1: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  heart_2020_df = pd.read_csv(FINAL_FILE_WITH_NANS_PATH)


In [ ]:
heart_2020_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401958 entries, 0 to 401957
Data columns (total 36 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   State                      401958 non-null  object 
 1   Sex                        401958 non-null  object 
 2   GeneralHealth              400997 non-null  object 
 3   PhysicalHealthDays         393267 non-null  float64
 4   MentalHealthDays           394029 non-null  float64
 5   LastCheckupTime            394851 non-null  object 
 6   PhysicalActivities         401276 non-null  object 
 7   SleepHours                 397264 non-null  float64
 8   RemovedTeeth               392161 non-null  object 
 9   HadHeartAttack             399875 non-null  object 
 10  HadAngina                  398656 non-null  object 
 11  HadStroke                  400772 non-null  object 
 12  HadAsthma                  400591 non-null  object 
 13  HadSkinCancer              40

In [ ]:
heart_2020_df.head()

,State,Sex,GeneralHealth,PhysicalHealthDays,MentalHealthDays,LastCheckupTime,PhysicalActivities,SleepHours,RemovedTeeth,HadHeartAttack,...,AgeCategory,HeightInMeters,WeightInKilograms,BMI,AlcoholDrinkers,HIVTesting,FluVaxLast12,PneumoVaxEver,CancerSurvivor,Year
0,Alabama,Female,Very good,3.0,30.0,5 or more years ago,Yes,5.0,"6 or more, but not all",No,...,Age 55 to 59,1.70,48.08,16.60,No,Yes,No,No,Yes,2020
1,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,1 to 5,No,...,Age 65 to 69,1.63,77.11,29.18,NaN,NaN,NaN,NaN,No,2020
2,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,NaN,No,...,Age 65 to 69,1.73,NaN,NaN,No,No,Yes,No,Yes,2020
3,Alabama,Female,Excellent,0.0,0.0,Within past 2 years (1 year but less than 2 ye...,No,6.0,1 to 5,No,...,Age 80 or older,NaN,NaN,NaN,No,No,No,No,No,2020
4,Alabama,Female,Very good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,None of them,No,...,Age 80 or older,1.68,57.15,20.34,No,NaN,Yes,Yes,No,2020


## **Estadísticas y valores únicos**

In [ ]:
heart_2020_df.describe(include='all')

,State,Sex,GeneralHealth,PhysicalHealthDays,MentalHealthDays,LastCheckupTime,PhysicalActivities,SleepHours,RemovedTeeth,HadHeartAttack,...,AgeCategory,HeightInMeters,WeightInKilograms,BMI,AlcoholDrinkers,HIVTesting,FluVaxLast12,PneumoVaxEver,CancerSurvivor,Year
count,401958,401958,400997,393267.000000,394029.000000,394851,401276,397264.000000,392161,399875,...,393701,380227.000000,367010.000000,360601.000000,375183,349091,376216,345526,400775,401958.0
unique,53,2,5,NaN,NaN,4,2,NaN,4,2,...,13,NaN,NaN,NaN,2,2,2,2,2,NaN
top,Minnesota,Female,Very good,NaN,NaN,Within past year (anytime less than 12 months ...,Yes,NaN,None of them,No,...,Age 65 to 69,NaN,NaN,NaN,Yes,No,Yes,No,No,NaN
freq,15835,218016,138139,NaN,NaN,311146,305883,NaN,209239,377918,...,41570,NaN,NaN,NaN,191565,232101,197197,205331,363989,NaN
mean,NaN,NaN,NaN,3.457595,3.916334,NaN,NaN,7.097099,NaN,NaN,...,NaN,1.701243,82.320386,28.306313,NaN,NaN,NaN,NaN,NaN,2020.0
std,NaN,NaN,NaN,8.081710,8.045309,NaN,NaN,1.468063,NaN,NaN,...,NaN,0.107235,20.994057,6.381649,NaN,NaN,NaN,NaN,NaN,0.0
min,NaN,NaN,NaN,0.000000,0.000000,NaN,NaN,1.000000,NaN,NaN,...,NaN,0.910000,22.680000,12.020000,NaN,NaN,NaN,NaN,NaN,2020.0
25%,NaN,NaN,NaN,0.000000,0.000000,NaN,NaN,6.000000,NaN,NaN,...,NaN,1.630000,68.040000,23.990000,NaN,NaN,NaN,NaN,NaN,2020.0
50%,NaN,NaN,NaN,0.000000,0.000000,NaN,NaN,7.000000,NaN,NaN,...,NaN,1.700000,79.380000,27.320000,NaN,NaN,NaN,NaN,NaN,2020.0
75%,NaN,NaN,NaN,2.000000,3.000000,NaN,NaN,8.000000,NaN,NaN,...,NaN,1.780000,92.990000,31.380000,NaN,NaN,NaN,NaN,NaN,2020.0


**Valores unicos**

In [ ]:
heart_2020_df.nunique().sort_values()

,0
Year,1
Sex,2
HadSkinCancer,2
PhysicalActivities,2
HadAsthma,2
HadStroke,2
HadAngina,2
HadCOPD,2
AlcoholDrinkers,2
HIVTesting,2


In [ ]:
# Variables categoricas
heart_2020_df.select_dtypes(include='object').columns

Index(['State', 'Sex', 'GeneralHealth', 'LastCheckupTime',
       'PhysicalActivities', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'TypeOfCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DifficultyWalking', 'DifficultyDressingBathing',
       'DifficultyErrands', 'SmokerStatus', 'ECigaretteUsage',
       'RaceEthnicityCategory', 'AgeCategory', 'AlcoholDrinkers', 'HIVTesting',
       'FluVaxLast12', 'PneumoVaxEver', 'CancerSurvivor'],
      dtype='object')

In [ ]:
# Variables numericas
heart_2020_df.select_dtypes(exclude='object').columns

Index(['PhysicalHealthDays', 'MentalHealthDays', 'SleepHours',
       'HeightInMeters', 'WeightInKilograms', 'BMI', 'Year'],
      dtype='object')

# **Valores faltantes**

In [ ]:
# Porcentaje de valores faltantes
missing = heart_2020_df.isnull().mean() * 100
missing.sort_values(ascending=False).head(20)

,0
TypeOfCancer,97.764194
ECigaretteUsage,87.395200
PneumoVaxEver,14.039278
HIVTesting,13.152369
BMI,10.288886
WeightInKilograms,8.694441
AlcoholDrinkers,6.661144
FluVaxLast12,6.404152
HeightInMeters,5.406286
SmokerStatus,5.249305


In [ ]:
# Tabla ordenada
missing_table = missing.reset_index()
missing_table.columns = ['Variable', 'Porcentaje_NaN']
missing_table = missing_table.sort_values('Porcentaje_NaN', ascending=False)
missing_table

,Variable,Porcentaje_NaN
14,TypeOfCancer,97.764194
24,ECigaretteUsage,87.395200
33,PneumoVaxEver,14.039278
31,HIVTesting,13.152369
29,BMI,10.288886
28,WeightInKilograms,8.694441
30,AlcoholDrinkers,6.661144
32,FluVaxLast12,6.404152
27,HeightInMeters,5.406286
23,SmokerStatus,5.249305


## **Limpieza de variables mayores a 40% de NaN**

In [ ]:
# Columnas que se mantienen
columns_to_keep = missing[missing <= 40].index
columns_to_keep

Index(['State', 'Sex', 'GeneralHealth', 'PhysicalHealthDays',
       'MentalHealthDays', 'LastCheckupTime', 'PhysicalActivities',
       'SleepHours', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DifficultyWalking', 'DifficultyDressingBathing',
       'DifficultyErrands', 'SmokerStatus', 'RaceEthnicityCategory',
       'AgeCategory', 'HeightInMeters', 'WeightInKilograms', 'BMI',
       'AlcoholDrinkers', 'HIVTesting', 'FluVaxLast12', 'PneumoVaxEver',
       'CancerSurvivor', 'Year'],
      dtype='object')

In [ ]:
heart_clean = heart_2020_df[columns_to_keep]
heart_clean.head()

,State,Sex,GeneralHealth,PhysicalHealthDays,MentalHealthDays,LastCheckupTime,PhysicalActivities,SleepHours,RemovedTeeth,HadHeartAttack,...,AgeCategory,HeightInMeters,WeightInKilograms,BMI,AlcoholDrinkers,HIVTesting,FluVaxLast12,PneumoVaxEver,CancerSurvivor,Year
0,Alabama,Female,Very good,3.0,30.0,5 or more years ago,Yes,5.0,"6 or more, but not all",No,...,Age 55 to 59,1.70,48.08,16.60,No,Yes,No,No,Yes,2020
1,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,1 to 5,No,...,Age 65 to 69,1.63,77.11,29.18,NaN,NaN,NaN,NaN,No,2020
2,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,NaN,No,...,Age 65 to 69,1.73,NaN,NaN,No,No,Yes,No,Yes,2020
3,Alabama,Female,Excellent,0.0,0.0,Within past 2 years (1 year but less than 2 ye...,No,6.0,1 to 5,No,...,Age 80 or older,NaN,NaN,NaN,No,No,No,No,No,2020
4,Alabama,Female,Very good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,None of them,No,...,Age 80 or older,1.68,57.15,20.34,No,NaN,Yes,Yes,No,2020


In [ ]:
# Resultado
print(f"Columnas originales: {heart_2020_df.shape[1]}")
print(f"Columnas después de limpieza: {heart_clean.shape[1]}")

Columnas originales: 36
Columnas después de limpieza: 34


In [ ]:
# Columnas eliminadas
columns_removed = missing[missing > 40].index
print("\nColumnas eliminadas por exceso de NaN (>40%):")
print(columns_removed.tolist())


Columnas eliminadas por exceso de NaN (>40%):
['TypeOfCancer', 'ECigaretteUsage']


# **Análisis de consistencia de variables**

## **Variables clave para el estudio**

In [ ]:
key_vars = ['HadHeartAttack','BMI','Sex','AgeCategory','SmokerStatus','RaceEthnicityCategory']

In [ ]:
# Copia del dataframe original
df_clean = heart_clean.copy()

### **Variable HadHeartAttack (Porcentaje de valores NaN -> 0.52%)**


In [ ]:
print("Valores faltantes antes de limpieza (HadHeartAttack):", df_clean['HadHeartAttack'].isna().sum())

Valores faltantes antes de limpieza (HadHeartAttack): 2083


In [ ]:
# Eliminar filas donde HadHeartAttack está vacío
df_clean = df_clean.dropna(subset=['HadHeartAttack'])

In [ ]:
df_clean['HadHeartAttack'].value_counts()

,count
HadHeartAttack,
No,377918
Yes,21957


In [ ]:
print(f"Cantidad de valores nulos en 'HadHeartAttack': {df_clean['HadHeartAttack'].isnull().sum()}")

Cantidad de valores nulos en 'HadHeartAttack': 0


### **Variable BMI (Porcentaje de valores NaN -> 10.29%)**

In [ ]:
print("Valores faltantes antes de imputación (BMI):", df_clean['BMI'].isna().sum())

Valores faltantes antes de imputación (BMI): 40912


In [ ]:
# Aplicamos la imputación con la mediana
bmi_median = df_clean['BMI'].median()
df_clean['BMI'].fillna(bmi_median, inplace=True)

/tmp/ipython-input-969017341.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean['BMI'].fillna(bmi_median, inplace=True)


In [ ]:
# Verificamos nuevamente los valores faltantes
print("Valores faltantes después de imputación (BMI):", df_clean['BMI'].isna().sum())
print("Mediana utilizada para imputación:", round(bmi_median, 2))

Valores faltantes después de imputación (BMI): 0
Mediana utilizada para imputación: 27.32


### **Variable Sex (Porcentaje de valores NaN -> 0%)**

### **Variable AgeCategory (Porcentaje de valores NaN -> 2.05%)**

In [ ]:
# Mostramos cuántos valores faltantes hay antes de la imputación
print("Valores faltantes antes de imputación:")
print(df_clean['AgeCategory'].isna().sum())

Valores faltantes antes de imputación:
8036


In [ ]:
# Aplicamos imputación con la moda (valor más frecuente)
for col in ['AgeCategory']:
    mode_val = df_clean[col].mode()[0]
    df_clean[col].fillna(mode_val, inplace=True)
    print(f"→ Variable '{col}' imputada con la moda: '{mode_val}'")

→ Variable 'AgeCategory' imputada con la moda: 'Age 65 to 69'


/tmp/ipython-input-3354019741.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(mode_val, inplace=True)


In [ ]:
df_clean['AgeCategory'].info()

<class 'pandas.core.series.Series'>
Index: 399875 entries, 0 to 401957
Series name: AgeCategory
Non-Null Count   Dtype 
--------------   ----- 
399875 non-null  object
dtypes: object(1)
memory usage: 6.1+ MB


In [ ]:
# Verificamos nuevamente los valores faltantes después de imputar
print("\nValores faltantes después de imputación:")
print(df_clean[['AgeCategory']].isna().sum())


Valores faltantes después de imputación:
AgeCategory    0
dtype: int64


In [ ]:
# Verificar que ya no haya valores faltantes en las variables clave
print(df_clean[['HadHeartAttack', 'Sex', 'BMI', 'SmokerStatus', 'RaceEthnicityCategory', 'AgeCategory']].isnull().mean() * 100)

HadHeartAttack           0.000000
Sex                      0.000000
BMI                      0.000000
SmokerStatus             5.207627
RaceEthnicityCategory    2.196937
AgeCategory              0.000000
dtype: float64


### **Variable SmokerStatus (Porcentaje de valores NaN -> 5.25%)**

In [ ]:
# Mostramos cuántos valores faltantes hay antes de la imputación
print("Valores faltantes antes de imputación:")
print(df_clean['SmokerStatus'].isna().sum())

Valores faltantes antes de imputación:
20824


In [ ]:
# Aplicamos imputación con la moda (valor más frecuente)
for col in ['SmokerStatus']:
    mode_val = df_clean[col].mode()[0]
    df_clean[col].fillna(mode_val, inplace=True)
    print(f"→ Variable '{col}' imputada con la moda: '{mode_val}'")

→ Variable 'SmokerStatus' imputada con la moda: 'Never smoked'


/tmp/ipython-input-3349073259.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(mode_val, inplace=True)


In [ ]:
df_clean['SmokerStatus'].value_counts()

,count
SmokerStatus,
Never smoked,244339
Former smoker,103319
Current smoker - now smokes every day,37962
Current smoker - now smokes some days,14255


In [ ]:
# Verificamos nuevamente los valores faltantes después de imputar
print("\nValores faltantes después de imputación:")
print(df_clean[['SmokerStatus']].isna().sum())


Valores faltantes después de imputación:
SmokerStatus    0
dtype: int64


### **Variable RaceEthnicityCategory (Porcentaje de valores NaN -> 2.24%)**

In [ ]:
# Mostramos cuántos valores faltantes hay antes de la imputación
print("Valores faltantes antes de imputación:")
print(df_clean['RaceEthnicityCategory'].isna().sum())

Valores faltantes antes de imputación:
8785


In [ ]:
# Aplicamos imputación con la moda (valor más frecuente)
for col in ['RaceEthnicityCategory']:
    mode_val = df_clean[col].mode()[0]
    df_clean[col].fillna(mode_val, inplace=True)
    print(f"→ Variable '{col}' imputada con la moda: '{mode_val}'")

→ Variable 'RaceEthnicityCategory' imputada con la moda: 'White only, Non-Hispanic'


/tmp/ipython-input-607064637.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(mode_val, inplace=True)


In [ ]:
df_clean['RaceEthnicityCategory'].value_counts()

,count
RaceEthnicityCategory,
"White only, Non-Hispanic",303521
Hispanic,36085
"Black only, Non-Hispanic",29918
"Other race only, Non-Hispanic",22086
"Multiracial, Non-Hispanic",8265


In [ ]:
# Verificamos nuevamente los valores faltantes después de imputar
print("\nValores faltantes después de imputación:")
print(df_clean[['SmokerStatus']].isna().sum())


Valores faltantes después de imputación:
SmokerStatus    0
dtype: int64


## **Imputación de variables restantes**

### **Variables completas**
*   **Variable State (Porcentaje de valores NaN -> 0%)**
*   **Variable Year (Porcentaje de valores NaN -> 0%)**

### **Imputación de variables numéricas**

*   **Variable SleepHours (Porcentaje de valores NaN -> 1.17%)**
*   **Variable PhysicalHealthDays (Porcentaje de valores NaN -> 2.16%)**
*   **Variable MentalHealthDays (Porcentaje de valores NaN -> 1.97%)**
*   **Variable HeightInMeters (Porcentaje de valores NaN -> 5.41%)**
*   **Variable WeightInKilograms (Porcentaje de valores NaN -> 8.69%)**


In [ ]:
# Imputación de variables numéricas
numeric_cols = ['SleepHours', 'PhysicalHealthDays', 'MentalHealthDays', 'HeightInMeters', 'WeightInKilograms']

In [ ]:
for col in numeric_cols:
    median_val = df_clean[col].median()
    df_clean[col].fillna(median_val, inplace=True)
    print(f"{col}: valores imputados con mediana {median_val:.2f}")

SleepHours: valores imputados con mediana 7.00
PhysicalHealthDays: valores imputados con mediana 0.00
MentalHealthDays: valores imputados con mediana 0.00
HeightInMeters: valores imputados con mediana 1.70
WeightInKilograms: valores imputados con mediana 79.38


/tmp/ipython-input-1228349538.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(median_val, inplace=True)


In [ ]:
# Verificar que ya no haya valores faltantes en las variables clave
print(df_clean[['SleepHours', 'PhysicalHealthDays', 'MentalHealthDays', 'HeightInMeters', 'WeightInKilograms']].isnull().mean() * 100)

SleepHours            0.0
PhysicalHealthDays    0.0
MentalHealthDays      0.0
HeightInMeters        0.0
WeightInKilograms     0.0
dtype: float64


### **Imputación de variables categóricas**

*   **Variable GeneralHealth (Porcentaje de valores NaN -> 0.24%)**
*   **Variable LastCheckupTime (Porcentaje de valores NaN -> 1.77%)**
*   **Variable RemovedTeeth (Porcentaje de valores NaN -> 2.44%)**

In [ ]:
# Imputación de variables categóricas ordinales
ordinal_cols = ['GeneralHealth', 'LastCheckupTime', 'RemovedTeeth']

In [ ]:
for col in ordinal_cols:
    mode_val = df_clean[col].mode()[0]
    df_clean[col].fillna(mode_val, inplace=True)
    print(f"{col}: valores imputados con moda '{mode_val}'")

GeneralHealth: valores imputados con moda 'Very good'
LastCheckupTime: valores imputados con moda 'Within past year (anytime less than 12 months ago)'
RemovedTeeth: valores imputados con moda 'None of them'


/tmp/ipython-input-333117031.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(mode_val, inplace=True)


In [ ]:
# Verificar que ya no haya valores faltantes en las variables clave
print(df_clean[['GeneralHealth', 'LastCheckupTime', 'RemovedTeeth']].isnull().mean() * 100)

GeneralHealth      0.0
LastCheckupTime    0.0
RemovedTeeth       0.0
dtype: float64


### **Imputación de variables binarias**

*   **Variable PneumoVaxEver (Porcentaje de valores NaN -> 14.04%)**
*   **Variable HIVTesting (Porcentaje de valores NaN -> 13.15%)**
*   **Variable AlcoholDrinkers (Porcentaje de valores NaN -> 6.66%)**
*   **Variable FluVaxLast12 (Porcentaje de valores NaN -> 6.40%)**
*   **Variable DifficultyErrands (Porcentaje de valores NaN -> 4.51%)**
*   **Variable DifficultyWalking (Porcentaje de valores NaN -> 4.17%)**
*   **Variable DifficultyDressingBathing (Porcentaje de valores NaN -> 4.13%)**
*   **Variable HadAngina (Porcentaje de valores NaN -> 0.82%)**
*   **Variable HadDepressiveDisorder (Porcentaje de valores NaN -> 0.57%)**
*   **Variable HadKidneyDisease (Porcentaje de valores NaN -> 0.52%)**
*   **Variable HadCOPD (Porcentaje de valores NaN -> 0.49%)**
*   **Variable HadArthritis (Porcentaje de valores NaN -> 0.38)**
*   **Variable HadAsthma (Porcentaje de valores NaN -> 0.34%)**
*   **Variable HadSkinCancer (Porcentaje de valores NaN -> 0.32%)**
*   **Variable HadStroke (Porcentaje de valores NaN -> 0.30%)**
*   **Variable CancerSurvivor (Porcentaje de valores NaN -> 0.29%)**
*   **Variable HadDiabetes (Porcentaje de valores NaN -> 0.20%)**
*   **Variable PhysicalActivities (Porcentaje de valores NaN -> 0.17%)**

In [ ]:
# Imputación de variables binarias Sí/No
binary_cols = ['PneumoVaxEver', 'HIVTesting', 'AlcoholDrinkers', 'FluVaxLast12', 'DifficultyErrands', 'DifficultyWalking',
               'DifficultyDressingBathing','HadAngina', 'HadDepressiveDisorder','HadKidneyDisease','HadCOPD','HadArthritis','HadAsthma','HadSkinCancer',
               'HadStroke','CancerSurvivor','HadDiabetes','PhysicalActivities']

In [ ]:
for col in binary_cols:
    mode_val = df_clean[col].mode()[0]
    df_clean[col].fillna(mode_val, inplace=True)
    print(f"{col}: valores imputados con moda '{mode_val}'")

/tmp/ipython-input-1213171526.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(mode_val, inplace=True)


PneumoVaxEver: valores imputados con moda 'No'
HIVTesting: valores imputados con moda 'No'
AlcoholDrinkers: valores imputados con moda 'Yes'
FluVaxLast12: valores imputados con moda 'Yes'
DifficultyErrands: valores imputados con moda 'No'
DifficultyWalking: valores imputados con moda 'No'
DifficultyDressingBathing: valores imputados con moda 'No'
HadAngina: valores imputados con moda 'No'
HadDepressiveDisorder: valores imputados con moda 'No'
HadKidneyDisease: valores imputados con moda 'No'
HadCOPD: valores imputados con moda 'No'
HadArthritis: valores imputados con moda 'No'
HadAsthma: valores imputados con moda 'No'
HadSkinCancer: valores imputados con moda 'No'
HadStroke: valores imputados con moda 'No'
CancerSurvivor: valores imputados con moda 'No'
HadDiabetes: valores imputados con moda 'No'
PhysicalActivities: valores imputados con moda 'Yes'


In [ ]:
# Verificar que ya no haya valores faltantes en las variables clave
print(df_clean[['PneumoVaxEver', 'HIVTesting', 'AlcoholDrinkers', 'FluVaxLast12', 'DifficultyErrands', 'DifficultyWalking',
               'DifficultyDressingBathing','HadAngina', 'HadDepressiveDisorder','HadKidneyDisease','HadCOPD','HadArthritis','HadAsthma','HadSkinCancer',
               'HadStroke','CancerSurvivor','HadDiabetes','PhysicalActivities','HadHeartAttack']].isnull().mean() * 100)

PneumoVaxEver                0.0
HIVTesting                   0.0
AlcoholDrinkers              0.0
FluVaxLast12                 0.0
DifficultyErrands            0.0
DifficultyWalking            0.0
DifficultyDressingBathing    0.0
HadAngina                    0.0
HadDepressiveDisorder        0.0
HadKidneyDisease             0.0
HadCOPD                      0.0
HadArthritis                 0.0
HadAsthma                    0.0
HadSkinCancer                0.0
HadStroke                    0.0
CancerSurvivor               0.0
HadDiabetes                  0.0
PhysicalActivities           0.0
HadHeartAttack               0.0
dtype: float64


## **Tratamiento de outliers (Valores extremos)**

In [ ]:
for var in ['BMI', 'SleepHours', 'HeightInMeters', 'WeightInKilograms']:
    Q1 = df_clean[var].quantile(0.25)
    Q3 = df_clean[var].quantile(0.75)
    IQR = Q3 - Q1
    df_clean = df_clean[(df_clean[var] >= Q1 - 1.5 * IQR) & (df_clean[var] <= Q3 + 1.5 * IQR)]

In [ ]:
print("\nValidación del DataFrame limpio:")
print(df_clean.info())


Validación del DataFrame limpio:
<class 'pandas.core.frame.DataFrame'>
Index: 370971 entries, 0 to 401957
Data columns (total 34 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   State                      370971 non-null  object 
 1   Sex                        370971 non-null  object 
 2   GeneralHealth              370971 non-null  object 
 3   PhysicalHealthDays         370971 non-null  float64
 4   MentalHealthDays           370971 non-null  float64
 5   LastCheckupTime            370971 non-null  object 
 6   PhysicalActivities         370971 non-null  object 
 7   SleepHours                 370971 non-null  float64
 8   RemovedTeeth               370971 non-null  object 
 9   HadHeartAttack             370971 non-null  object 
 10  HadAngina                  370971 non-null  object 
 11  HadStroke                  370971 non-null  object 
 12  HadAsthma                  370971 non-null  object 
 13  

In [ ]:
df_clean.head()

,State,Sex,GeneralHealth,PhysicalHealthDays,MentalHealthDays,LastCheckupTime,PhysicalActivities,SleepHours,RemovedTeeth,HadHeartAttack,...,AgeCategory,HeightInMeters,WeightInKilograms,BMI,AlcoholDrinkers,HIVTesting,FluVaxLast12,PneumoVaxEver,CancerSurvivor,Year
0,Alabama,Female,Very good,3.0,30.0,5 or more years ago,Yes,5.0,"6 or more, but not all",No,...,Age 55 to 59,1.70,48.08,16.60,No,Yes,No,No,Yes,2020
1,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,1 to 5,No,...,Age 65 to 69,1.63,77.11,29.18,Yes,No,Yes,No,No,2020
2,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,None of them,No,...,Age 65 to 69,1.73,79.38,27.32,No,No,Yes,No,Yes,2020
3,Alabama,Female,Excellent,0.0,0.0,Within past 2 years (1 year but less than 2 ye...,No,6.0,1 to 5,No,...,Age 80 or older,1.70,79.38,27.32,No,No,No,No,No,2020
4,Alabama,Female,Very good,0.0,0.0,Within past year (anytime less than 12 months ...,Yes,7.0,None of them,No,...,Age 80 or older,1.68,57.15,20.34,No,No,Yes,Yes,No,2020


## **Guardar archivo csv - Para Análisis Descriptivo**

In [ ]:
# Ruta a tu carpeta del proyecto
BASE = Path('/content/drive/MyDrive/Proyecto_de_tesis/datos_2020')
CLEAN = BASE / 'data' / 'cleaned'

In [ ]:
# Guardar el archivo limpio
output_path = CLEAN / 'heart_data_clean_2020_descriptive.csv'
df_clean.to_csv(output_path, index=False)

## **Normalización y codificación binaria**

In [ ]:
# Copia de base de datos para modelo predictivo
df_clean_pred = df_clean.copy()

In [ ]:
binary_vars = ['PneumoVaxEver', 'HIVTesting', 'AlcoholDrinkers', 'FluVaxLast12', 'DifficultyErrands', 'DifficultyWalking',
               'DifficultyDressingBathing','HadAngina', 'HadDepressiveDisorder','HadKidneyDisease','HadCOPD','HadArthritis','HadAsthma','HadSkinCancer',
               'HadStroke','CancerSurvivor','HadDiabetes','PhysicalActivities','HadHeartAttack']

In [ ]:
for col in binary_vars:
    df_clean_pred[col] = df_clean_pred[col].replace({'Yes': 1, 'No': 0})

/tmp/ipython-input-421723925.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_clean_pred[col] = df_clean_pred[col].replace({'Yes': 1, 'No': 0})


In [ ]:
def describe_df(df: pd.DataFrame):
    print(f"The dataset contains {df.shape[1]} columns and {len(df)} rows")

    for col in df.columns:
        col_dtype = df[col].dtype

        print(f"\nColumn: {col} ({col_dtype})")
        if col_dtype == 'object':
            print(f"--- Percentage of NaNs: {df[col].isna().sum() / len(df[col]) * 100}")
            print(f"--- Unique values:\n {df[col].unique()}")
        else:
            print(f"--- Summary statistics:\n {df[col].describe()}")

In [ ]:
describe_df(df_clean_pred)

The dataset contains 34 columns and 370971 rows

Column: State (object)
--- Percentage of NaNs: 0.0
--- Unique values:
 ['Alabama' 'Alaska' 'Arizona' 'Arkansas' 'California' 'Colorado'
 'Connecticut' 'Delaware' 'District of Columbia' 'Florida' 'Georgia'
 'Hawaii' 'Idaho' 'Illinois' 'Indiana' 'Iowa' 'Kansas' 'Kentucky'
 'Louisiana' 'Maine' 'Maryland' 'Massachusetts' 'Michigan' 'Minnesota'
 'Mississippi' 'Missouri' 'Montana' 'Nebraska' 'Nevada' 'New Hampshire'
 'New Jersey' 'New Mexico' 'New York' 'North Carolina' 'North Dakota'
 'Ohio' 'Oklahoma' 'Oregon' 'Pennsylvania' 'Rhode Island' 'South Carolina'
 'South Dakota' 'Tennessee' 'Texas' 'Utah' 'Vermont' 'Virginia'
 'Washington' 'West Virginia' 'Wisconsin' 'Wyoming' 'Guam' 'Puerto Rico']

Column: Sex (object)
--- Percentage of NaNs: 0.0
--- Unique values:
 ['Female' 'Male']

Column: GeneralHealth (object)
--- Percentage of NaNs: 0.0
--- Unique values:
 ['Very good' 'Good' 'Excellent' 'Fair' 'Poor']

Column: PhysicalHealthDays (float64)
-

In [ ]:
df_clean_pred.head()

,State,Sex,GeneralHealth,PhysicalHealthDays,MentalHealthDays,LastCheckupTime,PhysicalActivities,SleepHours,RemovedTeeth,HadHeartAttack,...,AgeCategory,HeightInMeters,WeightInKilograms,BMI,AlcoholDrinkers,HIVTesting,FluVaxLast12,PneumoVaxEver,CancerSurvivor,Year
0,Alabama,Female,Very good,3.0,30.0,5 or more years ago,1,5.0,"6 or more, but not all",0,...,Age 55 to 59,1.70,48.08,16.60,0,1,0,0,1,2020
1,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,1,7.0,1 to 5,0,...,Age 65 to 69,1.63,77.11,29.18,1,0,1,0,0,2020
2,Alabama,Female,Good,0.0,0.0,Within past year (anytime less than 12 months ...,1,7.0,None of them,0,...,Age 65 to 69,1.73,79.38,27.32,0,0,1,0,1,2020
3,Alabama,Female,Excellent,0.0,0.0,Within past 2 years (1 year but less than 2 ye...,0,6.0,1 to 5,0,...,Age 80 or older,1.70,79.38,27.32,0,0,0,0,0,2020
4,Alabama,Female,Very good,0.0,0.0,Within past year (anytime less than 12 months ...,1,7.0,None of them,0,...,Age 80 or older,1.68,57.15,20.34,0,0,1,1,0,2020


## **Guardar archivo csv - Para Modelo Predictivo**

In [ ]:
# Ruta a tu carpeta del proyecto
BASE = Path('/content/drive/MyDrive/Proyecto_de_tesis/datos_2020')
CLEAN = BASE / 'data' / 'cleaned'

In [ ]:
# Guardar el archivo limpio
output_path = CLEAN / 'heart_data_clean_2020_predictive.csv'
df_clean_pred.to_csv(output_path, index=False)